File: data/scripts/seed_search_intent_model.mjs
#!/usr/bin/env node
// Atomically publish only the trained intent graph. This deliberately reads
// Mongo settings from the root .env and never loads AI/image provider keys.

import { existsSync, readFileSync } from "node:fs";
import { resolve } from "node:path";
import { MongoClient } from "mongodb";

const projectRoot = resolve(import.meta.dirname, "../..");
const modelPath = resolve(projectRoot, "data/processed/seed/search_intent_model.json");
const envPath = resolve(projectRoot, ".env");

function readEnv(path) {
  const values = {};
  if (!existsSync(path)) return values;
  for (const raw of readFileSync(path, "utf8").split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const separator = line.indexOf("=");
    const key = line.slice(0, separator).trim();
    let value = line.slice(separator + 1).trim();
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) value = value.slice(1, -1);
    values[key] = value;
  }
  return values;
}

const fileEnv = readEnv(envPath);
const uri = process.env.MONGODB_URL || process.env.MONGODB_URI || fileEnv.MONGODB_URL || fileEnv.MONGODB_URI;
const databaseName = process.env.MONGODB_DB_NAME || process.env.DATABASE_NAME || fileEnv.MONGODB_DB_NAME || fileEnv.DATABASE_NAME || "stylme";
if (!uri) throw new Error(`MONGODB_URL (or MONGODB_URI) is required in ${envPath}.`);
const model = JSON.parse(readFileSync(modelPath, "utf8"));
if (!model.key || !model.nodes || model.training_rows !== 30000) throw new Error("Intent model is incomplete; rebuild it before seeding.");

const client = new MongoClient(uri, { serverSelectionTimeoutMS: 15_000 });
try {
  await client.connect();
  const collection = client.db(databaseName).collection("search_intent_models");
  await collection.createIndex({ key: 1 }, { unique: true });
  await collection.replaceOne({ key: model.key }, model, { upsert: true });
  console.log(JSON.stringify({ status: "seeded", database: databaseName, key: model.key, version: model.version, trainingRows: model.training_rows, nodes: Object.keys(model.nodes).length }, null, 2));
} finally {
  await client.close();
}

File: data/scripts/seed_search_intent_model.py
#!/usr/bin/env python3
"""Upsert the trained, non-generative search graph into StylMe MongoDB."""

from __future__ import annotations

import asyncio
import json
import sys
from pathlib import Path


REPOSITORY_ROOT = Path(__file__).resolve().parents[2]
BACKEND_ROOT = REPOSITORY_ROOT / "backend"
MODEL_PATH = REPOSITORY_ROOT / "data" / "processed" / "seed" / "search_intent_model.json"
sys.path.insert(0, str(BACKEND_ROOT))

from app.database.connection import mongo_runtime  # noqa: E402


async def seed() -> None:
    model = json.loads(MODEL_PATH.read_text(encoding="utf-8"))
    if model.get("training_rows") != 30_000:
        raise RuntimeError("Search graph must be trained from exactly 30,000 products")
    database = await mongo_runtime.connect()
    try:
        await database.search_intent_models.replace_one(
            {"key": model["key"]},
            model,
            upsert=True,
        )
        await database.search_intent_models.create_index("key", unique=True)
        print(json.dumps({
            "key": model["key"],
            "trainingRows": model["training_rows"],
            "nodes": model["statistics"]["nodes"],
            "edges": model["statistics"]["edges"],
        }))
    finally:
        await mongo_runtime.close()


if __name__ == "__main__":
    asyncio.run(seed())
